codex/stores/sqlitestore.nim

```diff
@@ -0,0 +1,204 @@
+## Nim-Codex
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import pkg/upraises
+
+push: {.upraises: [].}
+
+import std/options
+
+import pkg/chronos
+import pkg/chronicles
+import pkg/datastore/sqlite_datastore
+import pkg/libp2p
+import pkg/questionable
+import pkg/questionable/results
+
+import ./blockstore
+import ./cachestore
+
+export blockstore, sqlite_datastore
+
+logScope:
+  topics = "codex sqlitestore"
+
+type
+  SQLiteStore* = ref object of BlockStore
+    cache: BlockStore
+    datastore: SQLiteDatastore
+
+const
+  allBlocks = Query.init(Key.init("*").expect("should not fail"))
+
+proc new*(
+  T: type SQLiteStore,
+  repoDir: string,
+  cache: BlockStore = CacheStore.new()): T =
+
+  T(cache: cache,
+    datastore: SQLiteDatastore.new(repoDir).expect("should not fail"))
+
+proc datastore*(self: SQLiteStore): SQLiteDatastore =
+  self.datastore
+
+proc blockKey*(blockCid: Cid): Key =
+  Key.init($blockCid).expect("should not fail")
+
+method getBlock*(
+  self: SQLiteStore,
+  cid: Cid): Future[?!(?Block)] {.async.} =
+  ## Get a block from the cache or database, save a copy to the cache
+  ##
+
+  trace "Getting block from database", cid
+  if cid.isEmpty:
+    trace "Empty block, ignoring"
+    return success cid.emptyBlock.some
+
+  let
+    getCachedRes = await self.cache.getBlock(cid)
+
+  if getCachedRes.isErr:
+    trace "Unable to read block from cache", cid, error = getCachedRes.error.msg
```

Comment:
I made this symmetric with the "trace but don't fail" logic of L95-L96.

For `FSStore` when there's an error reading from the cache it does `return failure`

https://github.com/status-im/nim-codex/blob/7cfb157d327624d848a09c2df4eda20ea677a8e4/codex/stores/fsstore.nim#L49-L51

And when there's an error writing to the cache it only traces

https://github.com/status-im/nim-codex/blob/7cfb157d327624d848a09c2df4eda20ea677a8e4/codex/stores/fsstore.nim#L101-L102

I think it would be best to be consistent re: errors when accessing the cache, but is it better to only trace or trace + `return failure`?